### PR TITLE
Ex.3 dev workshop commit

### PR DESCRIPTION
**Pull request by swarnadipa-dev**

### CODE REVIEW

**Review by Gabez0r — COMMENTED** [review]
Great job! 👏  But, are we testing all paths? We should test that if we click "Cancel" instead of "Ok", that nothing happens.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 1]
```diff
@@ -85,12 +98,25 @@ suite('nuxeo-favorites-toggle-button', () => {
...
 
-    test('toggle should add the document to favorites', async () => {
-      // Add the documents to favorites by toggling
+    test('it should display a confirm popup', () => {
+      const confirmSpy = sinon.stub(window, 'confirm');
+      tap(element);
+      expect(confirmSpy.called).to.be.true;
+    });
+
+    test('toggle should remove document from favorites when clicked ok', async () => {
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: We should also test what happens if the user clicks cancel.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 2]
```diff
@@ -48,12 +48,25 @@ suite('nuxeo-favorites-toggle-button', () => {
...
     });
 
-    test('toggle should remove document from favorites', async () => {
+    test('it should display a confirm popup', () => {
+      const confirmSpy = sinon.stub(window, 'confirm');
+      tap(element);
+      expect(confirmSpy.called).to.be.true;
+    });
+
+    test('toggle should remove document from favorites when clicked ok', async () => {
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js:65]: We should also test what happens if the user clicks cancel in this situation. We need to make sure that nothing happens.

Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js:65]: We should also test what happens if the user clicks cancel in this situation.

Comment (richardsd) [reply ui/test/nuxeo-favorites-toggle-button.test.js:65]: I would say to add a new test case for `toggle should not remove document from favorites when clicked cancel`

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 3]
```diff
@@ -84,16 +92,32 @@ suite('nuxeo-favorites-toggle-button', () => {
...
+    teardown(() => {
+      if (window.confirm.restore) {
+        window.confirm.restore();
+      }
+    });
+
+    test('toggle should not remove document from favorites when clicked Cancel', async () => {
+      sinon.stub(window, 'confirm').returns(false);
+      expect(element.favorite).to.be.true;
+    });
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: I think this is in the wrong suite. Note that this unit test is actually failing.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 4]
```diff
@@ -84,16 +92,32 @@ suite('nuxeo-favorites-toggle-button', () => {
...
     });
+    teardown(() => {
+      if (window.confirm.restore) {
+        window.confirm.restore();
+      }
+    });
+
+    test('toggle should not remove document from favorites when clicked Cancel', async () => {
+      sinon.stub(window, 'confirm').returns(false);
+      expect(element.favorite).to.be.true;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: We didn't click the button before the assertion, right? So why would it change?

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 5]
```diff
@@ -84,16 +92,32 @@ suite('nuxeo-favorites-toggle-button', () => {
...
+    });
+
+    test('toggle should not add the document to favorites when clicked on Cancel', async () => {
+      sinon.stub(window, 'confirm').returns(false);
+      expect(element.favorite).to.be.false;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: We didn't click the button before the assertion, right? So why would it change?

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 6]
```diff
@@ -34,8 +37,15 @@ suite('nuxeo-favorites-toggle-button', () => {
...
         },
+
         facets: [],
       };
+      button = await fixture(
+        html`
+          <nuxeo-favorites-toggle-button></nuxeo-favorites-toggle-button>
+        `,
+      );
+
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: I don't understand why we need a new button element.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 7]
```diff
@@ -48,16 +58,31 @@ suite('nuxeo-favorites-toggle-button', () => {
...
+      sinon.stub(window, 'confirm').returns(true);
       tap(element);
       await waitChanged(element, 'favorite');
       expect(element.favorite).to.be.false;
     });
+
+    test('toggle should not remove document from favorites when clicked Cancel', async () => {
+      sinon.spy(button.$.opRemove, 'execute');
+      sinon.stub(window, 'confirm').returns(false);
+      expect(button.$.opRemove.execute.notCalled).to.be.true;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: This is a good idea, to check that the function was not called.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 8]
```diff
@@ -48,16 +58,31 @@ suite('nuxeo-favorites-toggle-button', () => {
...
+      expect(element.favorite).to.be.true;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: The problem with this test remains. We did not interact with the element (i.e., tap/click on it), so we are not actually testing what happens when the user hits the cancel button.

File: ui/test/nuxeo-favorites-toggle-button.test.js [hunk 9]
```diff
@@ -84,16 +109,30 @@ suite('nuxeo-favorites-toggle-button', () => {
...
       tap(element);
       await waitChanged(element, 'favorite');
       expect(element.favorite).to.be.true;
     });
+
+    test('toggle should not add the document to favorites when clicked on Cancel', async () => {
+      sinon.spy(button.$.opAdd, 'execute');
+      sinon.stub(window, 'confirm').returns(false);
+      expect(button.$.opRemove.execute.notCalled).to.be.true;
+      expect(element.favorite).to.be.false;
```
Comment (Gabez0r) [inline ui/test/nuxeo-favorites-toggle-button.test.js]: This is also not testing much, as we are checking two different instances of the element, one of which doesn't even have a document.

**Review by Gabez0r — COMMENTED** [review]
well done!

**Review by richardsd — COMMENTED** [review]
Looks good to me 👍 Please don't forget that the Web UI cross-repo is failing (its the csv ftest). You can retry the workflow to run the ftests again.

**Pull request closed without merging**